### PR TITLE
[WIP] Fake bls at runtime

### PR DESCRIPTION
**Pull request by gnattishness**
Fake bls implementation using runtime `UpdateFlags`.

Quite basic and just skips bls_verify where it appears.

Alternative to #671

### CODE REVIEW

**Comment by gnattishness** [comment]
Current open questions:

- What to do for the few remaining instances of `skipValidation`?
  - Could have `skipValidation` as a "parent" flag (subset?) that implies all skipXXValidation flags are enabled
  - Could convert the existing `if skipValidation` to more specific flags e.g. `skipStateRootValidation`
- Is there a reasonable way to log a warning if skipBLSValidation is used? (without spamming)
- What extra (if anything) needs to be done for v0.10.x `FastAggregateVerify` etc operations

**Comment by gnattishness** [comment]
Currently fails the following tests, but don't appear to be caused by my changes:
```
[Suite] SyncManager test suite
  [OK] PeerSlot tests
  [OK] PeerGroup tests
  [OK] SyncQueue non-async tests
  [OK] SyncQueue async tests
  [OK] SyncManager one-peer test
  [OK] SyncManager one-peer-slot test
  [OK] SyncManager one-peer-group test
  [OK] SyncManager group-recovery test
ERR 2020-02-07 04:05:03+00:00 Synchronization failed                     tid=31026 duration=301ms523us577ns errors=3
  [OK] SyncManager failure test

[Suite] [Unit - Spec - Block processing] Deposits  [Preset: mainnet]
/eth2/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(39) sysFatal
Error: unhandled exception: /eth2/nim-beacon-chain/tests/spec_block_processing/test_process_deposits.nim(30, 12) `genesisState.validators.len == int NumValidators`  [AssertionError]
Error: execution of an external program failed: '/eth2/nim-beacon-chain/build/all_tests '
stack trace: (most recent call last)
/eth2/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/nimscript.nim(393, 18)
/eth2/nim-beacon-chain/beacon_chain.nimble(49, 15) testTask
/eth2/nim-beacon-chain/beacon_chain.nimble(44, 8) buildBinary
/eth2/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/nimscript.nim(252, 7) exec
/eth2/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/nimscript.nim(252, 7) Error: unhandled exception: FAILED: nim c --out:./build/all_tests -r -d:release -d:chronicles_log_level=ERROR -d:const_preset=mainnet --verbosity:0 --hints:off --warnings:off -d:usePcreHeader --passL:"-lpcre" tests/all_tests.nim  [OSError]
```

**Comment by arnetheduck** [comment]
we've planned splitting up `SkipValidation` into more specific checks, no need to make it a "parent" check - if needed, it's possible to introduce a `const SkipAllValidation = {SkipA, SkipB}`

**Comment by arnetheduck** [comment]
> reasonable way to log a warning

where do you want the warning?

we use the same flag to disable bls verification when replaying blocks from the database - basically, we only ever store blocks that have passed validation, so when we need to call up an arbitrary state, we load the most recent one before and replay.

**Comment by mratsim** [comment]
Can probably be done later but is there a reason to skip those at runtime instead of compile-time?

**Comment by gnattishness** [comment]
> > reasonable way to log a warning
> 
> where do you want the warning?
> 
> we use the same flag to disable bls verification when replaying blocks from the database - basically, we only ever store blocks that have passed validation, so when we need to call up an arbitrary state, we load the most recent one before and replay.

Ah interesting! I'm not aware of any other implementations that have that optimization yet (though could easily be wrong).
The warning was more a thought to protect people from accidentally using it in production, but that's less useful when you have a valid reason to use it!
Also not too likely for them to accidentally use it, unlike the compile-time flag which can be set via the environment.

**Comment by gnattishness** [comment]
> Can probably be done later but is there a reason to skip those at runtime instead of compile-time?

Skip what?

**Comment by mratsim** [comment]
In Nim you can enforce compile-time evaluation and compile-time parameters with the `static` modifier, for example:

https://github.com/status-im/nim-beacon-chain/blob/612f0f165d95ef464a4794f606c783222477ff61/beacon_chain/spec/state_transition_block.nim#L91-L125

Can be rewritten as:

```Nim
proc process_randao(
    state: var BeaconState, body: BeaconBlockBody, flags: static UpdateFlags,
    stateCache: var StateCache): bool {.nbench.}=
  # Notice the new static flag
  let
    epoch = state.get_current_epoch()
    proposer_index = get_beacon_proposer_index(state, stateCache)

  if proposer_index.isNone:
    debug "Proposer index missing, probably along with any active validators"
    return false

  let proposer = addr state.validators[proposer_index.get]

  # Verify that the provided randao value is valid
  # The `when` branch is evaluated at compile-time and completely disappear
  #  from the binary if invalid
  # i.e. skipping the checks at compile-time
  when skipBLSValidation notin flags:
    if not bls_verify(
      proposer.pubkey,
      hash_tree_root(epoch.uint64).data,
      body.randao_reveal,
      get_domain(state, DOMAIN_RANDAO)):

      notice "Randao mismatch", proposer_pubkey = proposer.pubkey,
                                message = epoch,
                                signature = body.randao_reveal,
                                slot = state.slot
      return false

  # Mix it in
  let
    mix = get_randao_mix(state, epoch)
    rr = eth2hash(body.randao_reveal.getBytes()).data

  for i in 0 ..< mix.data.len:
    state.randao_mixes[epoch mod EPOCHS_PER_HISTORICAL_VECTOR].data[i] = mix.data[i] xor rr[i]

  true
```

The `updateFlags` (which are in an internal proc) and the branch that test the flag presence completely disappear from the final binary. We can choose to make those flags a compile option with no runtime cost.

The tradeoff is that we have 2 different binaries, one for test and one for production, but AFAIK when building Nimbus for nfuzz we already require this distinction.

This question is also for @arnetheduck, @zah, @tersec and @stefantalpalaru.

**Comment by stefantalpalaru** [comment]
> 2 different binaries, one for test and one for production

There's a risk that some code will end up in production without being tested in CI?

**Comment by arnetheduck** [comment]
> I'm not aware of any other implementations that have that optimization yet

fwiw, it's an outcome of the architecture of having tainted/untainted "zones" more than a deliberate optimization

> given that in production this should never be done.

per earlier comments, we use the flags in "production" as well - BLS checks are done only once when receiving them from external sources, then they enter a "trusted area". skipping tree roots (the other thing `skip*` does) is done when producing blocks because the state tree hash is not yet computed. 

> We can choose to make those flags a compile option with no runtime cost.

in practice, making the flag `static` will mean at least two or three full copies of the state transition code because nim will expand it for every constant value passed into it, not counting the additional developer mental overhead of understanding the limitations and costs of `static`. there's no free lunch. there will not be two binaries, there will be a single binary with all variations of the code. `llvm` and `gcc` do this same optimization (constant specialization), but with a cost model. 

if I were to predict what the cost model will do, it would be that for the fuzzer that calls "simple" functions, it will do the specialization, but for the "production" code, it will not because that would involve duplicating the full `state_transition`.

`static` is useful when at compile time there are paths that become invalid due to type/memory size/other compile time considerations that would render the code inside `when` non-compileable under certain input values. 

there's also the rare case where the programmer knows better than the compiler cost model. `register` and `inline` where effectively deprecated in `c` because most often, the programmer doesn't.

on a more serious note, the wall clock time spent discussing whether this should be a feature from an optimization point of view is likely already more than we will ever recover from few `if`'s that are avoided.. the only focus here should be legibility and clarity and stability of the code - nothing else matters unless there's an accompanying benchmark that shows a clear benefit.

from a functional point of view as far as this PR goes, I'd love to see each "kind" of check skipped have its own flag (bls, state root, don't remember if there's anything else) - that would make it easy to look for in an audit and provide good readability.

**Comment by zah** [comment]
Personally, I'm quite happy with this PR, but in order to merge it, we need to do the following:

1) Rebase it
2) Review all usages of the current `skipValidation` flag (which used to mean "skip all validations") and turn them into a combination of the more specific flags that start to appear now. As the code stands right now, the old `skipValidation` won't skip the BLS validations controlled by the new flag.

**Comment by gnattishness** [comment]
> In Nim you can enforce compile-time evaluation and compile-time parameters with the `static` modifier, for example:
Thanks for the great explanation!

**Comment by gnattishness** [comment]
> the only focus here should be legibility and clarity and stability of the code - nothing else matters unless there's an accompanying benchmark that shows a clear benefit.

Good advice/approach. Sorry to trigger this distraction :)

> I'd love to see each "kind" of check skipped have its own flag (bls, state root, don't remember if there's anything else) - that would make it easy to look for in an audit and provide good readability.

> 1. Rebase it
> 1. Review all usages of the current skipValidation flag (which used to mean "skip all validations") and turn them into a combination of the more specific flags that start to appear now. As the code stands right now, the old skipValidation won't skip the BLS validations controlled by the new flag.

I'll sort these out and ping back. Only issue might be trying to figure out which should be applied in certain tests but I'll let you know.

File: research/state_sim.nim [hunk 1]
```diff
@@ -67,19 +68,20 @@ cli do(slots = SLOTS_PER_EPOCH * 6,
...
+    flags = if validate: {} else: {skipBlsValidation}
     deposits = makeInitialDeposits(validators, flags)
 
   echo "Generating Genesis..."
 
   let
     genesisState =
       initialize_beacon_state_from_eth1(
-        Eth2Digest(), 0, deposits, {skipMerkleValidation})
+        Eth2Digest(), 0, deposits, flags + {skipMerkleValidation})
```
Comment (gnattishness) [inline research/state_sim.nim]: Note: now forwarding the `skipBlsValidation` flag so consistent across the deposit creation and the processing.

File: tests/test_beacon_chain_db.nim [hunk 2]
```diff
@@ -104,7 +104,7 @@ suite "Beacon chain DB" & preset():
 
     let
       state = initialize_beacon_state_from_eth1(
-        eth1BlockHash, 0, makeInitialDeposits(SLOTS_PER_EPOCH), {skipValidation})
+        eth1BlockHash, 0, makeInitialDeposits(SLOTS_PER_EPOCH), {skipBlsValidation})
```
Comment (gnattishness) [inline tests/test_beacon_chain_db.nim]: Not sure if we want to be skipping merkle validation here too?

File: beacon_chain/spec/state_transition_block.nim [hunk 3]
```diff
@@ -56,14 +59,13 @@ proc process_block_header*(
     return false
 
   # Verify that the parent matches
-  if skipValidation notin flags and not (blck.parent_root ==
+  if skipBlockParentRootValidation notin flags and not (blck.parent_root ==
       hash_tree_root(state.latest_block_header)):
-    # TODO: skip validation is too strong
-    #       can't do "invalid_parent_root" test
```
Comment (gnattishness) [inline beacon_chain/spec/state_transition_block.nim]: Can't see the "invalid_parent_root" test being ignored currently.

**Comment by gnattishness** [comment]
@mratsim @zah Applied changes and removed the generic `skipValidation` flag.
Currently passes tests, but possible there are some unnecessary `skipXXXValidation` flags applied.

If this looks ok, let me know before you merge so I can squash.

File: tests/test_beacon_chain_db.nim (hunk 2 above)
Comment (arnetheduck) [reply tests/test_beacon_chain_db.nim]: probably, yes - it's not the objective of these tests, and it's tested elsewhere

Comment (gnattishness) [reply tests/test_beacon_chain_db.nim]: Done

**Pull request merged**